File: CPP/Trie.cpp
#include <cassert>
#include <unordered_map>

class TrieNode {
private:
    char val;
    std::unordered_map<char, TrieNode*> children;
    bool isEnd;
public:
    /**
     * Constructor
     */
    TrieNode(char v, bool end = false);

    /**
     * Creates a new TrieNode and inserts it as a child node to the current node.
     * If the child node already exists, does nothing.
     * Finally, it returns a pointer to the child node.
     */
    TrieNode* addChild(char c);

    /**
     * Sets the current TrieNode as leaf node, by setting 
     * it's `isEnd` attribute.
     */
    void setEnd();

    /**
     * Gets the value of `isEnd` attribute.
     */
    bool getEnd();

    /**
     * Returns a pointer to the child node if it exists, otherwise return `nullptr`.
     */
    TrieNode* hasChild(char c);
};

TrieNode::TrieNode(char v, bool end) : val(v), isEnd(end) {
}

TrieNode* TrieNode::addChild(char c) {
    TrieNode* t;
    if(this->children.count(c)) {
        t = this->children[c];
    } else {
        t = new TrieNode(c);
        // this->children[c].insert(std::make_pair<char, TrieNode*>(c, t));
        this->children[c] = t;
    }
    return t;
}

void TrieNode::setEnd() {
    this->isEnd = true;
}

bool TrieNode::getEnd() {
    return isEnd;
}

TrieNode* TrieNode::hasChild(char c) {
    TrieNode* t = nullptr;
    if(this->children.count(c)){
        t = this->children[c];
    }
    return t;
}

class Trie {
private:
    TrieNode* root;
public:
    /**
     * Default constructor
     */
    Trie();

    /**
     * Inserts the `std::string` into the Trie
     */
    void insert(std::string s);

    /**
     * Returns `true` if this `std::string` `s` exists in the Trie.
     * or, if the `std::string` `s` is present in the Trie as a substring 
     * of a longer string when `allowPrefix` is `true`.
     */
    bool search(std::string s, bool allowPrefix = false);
};

Trie::Trie(){
    root = new TrieNode(0);
}

void Trie::insert(std::string s) {
    TrieNode* t = root;
    for(char& a: s) {
        t = t->addChild(a);
    }
    t->setEnd();
}

bool Trie::search(std::string s, bool allowPrefix) {
    TrieNode* t = root;
    for(char& a: s) {
        t = t->hasChild(a);
        if(t == nullptr) {
            return false;
        }
    }
    return allowPrefix || t->getEnd();
}

int main() {
    Trie t;

    t.insert("abcd");
    t.insert("ab");
    assert(t.search("abc") == false);
    assert(t.search("abc", true) == true);
    assert(t.search("ab")  ==  true);
    assert(t.search("xyz") == false);
    assert(t.search("xyz", true) == false);

    return 0;
}

File: GoLang/Trie.go
package main

import "fmt"

type TrieNode struct {
	val      rune
	children map[rune]*TrieNode
	isEnd    bool
}

// NewTrieNode is used to construct an instance of TrieNode
func NewTrieNode(val rune) *TrieNode {
	tn := new(TrieNode)
	tn.val = val
	tn.children = make(map[rune]*TrieNode)
	tn.isEnd = false
	return tn
}

// AddChild creates a new TrieNode and inserts it as a
// child node to the current node.
// If the child node already exists, does nothing.
// Finally, it returns a pointer to the child node.
func (this *TrieNode) AddChild(c rune) *TrieNode {
	var t *TrieNode
	if val, ok := this.children[c]; ok {
		t = val
	} else {
		t = NewTrieNode(c)
		this.children[c] = t
	}
	return t
}

// HasChild returns a pointer to the child node if it exists,
// otherwise return `nil`.
func (this *TrieNode) HasChild(c rune) *TrieNode {
	var t *TrieNode = nil
	if val, ok := this.children[c]; ok {
		t = val
	}
	return t
}

// GetEnd gets the value of `isEnd` attribute.
func (this *TrieNode) GetEnd() bool {
	return this.isEnd
}

// SetEnd sets the current TrieNode as leaf node, by setting
// it's `isEnd` attribute.
func (this *TrieNode) SetEnd(b bool) {
	this.isEnd = b
}

type Trie struct {
	root *TrieNode
}

// NewTrie constructs an instance of Trie and returns a pointer to it
func NewTrie() *Trie {
	var t *Trie = new(Trie)
	t.root = NewTrieNode(0)
	return t
}

// Insert inserts the string into the Trie
func (this *Trie) Insert(s string) {
	var t *TrieNode = this.root
	for _, a := range s {
		t = t.AddChild(a)
	}
	t.SetEnd(true)
}

// Search returns `true` if the string `s` exists in the Trie.
// or, if the string `s` is present in the Trie as a substring
// of a longer string when `allowPrefix` is `true`.
func (this *Trie) Search(s string, allowPrefix bool) bool {
	var t *TrieNode = this.root
	for _, a := range s {
		t = t.HasChild(a)
		if t == nil {
			return false
		}
	}

	return allowPrefix || t.GetEnd()
}

// PanicIf works as an equivalent of the C++ assert()
func PanicIf(b bool) {
	if !b {
		panic(b)
	}
}

func main() {

	var t *Trie = NewTrie()
	t.Insert("abcd")
	t.Insert("ab")
	PanicIf(t.Search("abc", false) == false)
	PanicIf(t.Search("abc", true) == true)
	PanicIf(t.Search("ab", false) == true)
	PanicIf(t.Search("xyz", false) == false)
	PanicIf(t.Search("xyz", true) == false)

	fmt.Println("Trie works correctly")
}
